### PR TITLE
Add BaseFormatConverter::convertIcuShortFormatToPattern

### PR DESCRIPTION
**Pull request by nkovacs**
This removes code duplication and makes it easier to extend
FormatConverter with custom conversions.

### CODE REVIEW

**Comment by nkovacs** [comment]
Why is there a separate `$juiFallbackDatePatterns` and why doesn't it contain time information?
It would make more sense to have a single `$fallbackPatterns` array, which contains ICU patterns, and then that can be converted normally. That would make it simpler to create a custom conversion function, since you'd always get a valid ICU pattern, and you wouldn't have to deal with fallbacks.

**Comment by nkovacs** [comment]
Oh, I see, convertDateIcuToJui doesn't convert times. Why?

**Comment by cebe** [comment]
jui is about the jQueryUI date picker which can not pick times and thus has no syntax for time values.

**Comment by nkovacs** [comment]
I see. It would still be possible to use a single `$fallbackPatterns` array, and the conversion would strip out anything jui doesn't support. But because the patterns are public, this would break BC.

**Comment by cebe** [comment]
> I see. It would still be possible to use a single $fallbackPatterns array, and the conversion would strip out anything jui doesn't support. But because the patterns are public, this would break BC.

that would be possible in general but I think we should leave it as is. This way there is more control over what actually happens. This is a fallback for the case when intl is not installed anyway. If you want to do it properly you should have installed intl extension anyway.

File: framework/helpers/BaseFormatConverter.php [hunk 1]
```diff
@@ -81,6 +81,37 @@ class BaseFormatConverter
...
+     * Converts the ICU short patterns `short`, `medium`, `long` and `full` into a full ICU pattern.
+     *
+     * @param string $pattern an ICU short pattern.
+     * @param string $type 'date', 'time', or 'datetime'.
+     * @param string $locale the locale to use for converting the patterns.
+     * If not given, `Yii::$app->language` will be used.
+     * @return string|false The converted date format pattern, or false if the pattern could not be converted because
+     * the intl extension is not loaded. If the pattern is not an ICU short pattern, it will be returned unmodified.
+     */
+    public static function convertIcuShortFormatToPattern($pattern, $type = 'date', $locale = null)
```
Comment (cebe) [inline framework/helpers/BaseFormatConverter.php]: I'd prefer this to be protected instead. Would that still fit your use case?

Comment (cebe) [inline framework/helpers/BaseFormatConverter.php]: Also need an `@since 2.0.4` annotation.

**Comment by nkovacs** [comment]
It would be much simpler if convertIcuShortFormatToPattern could just return a pattern every time, and you wouldn't have to deal with separate fallbacks on a case-by-case basis.
See https://github.com/nkovacs/yii2-datetimepicker/blob/master/FormatConverter.php#L56

**Comment by cebe** [comment]
I know but that will break BC so we can not do that in 2.0.x. also currently the format is specified in the PHP / JUI syntax which is more explicit than having a syntax that needs to be converted after wards.

**Comment by nkovacs** [comment]
I don't think it's a good idea to have separate fallbacks, if they are different and you install intl, your app will suddenly use a different format and may break.
Since I had to reimplement this anyway for my extension, I can wait for Yii 2.1.

**Comment by cebe** [comment]
related to #11732

**Comment by samdark** [comment]
Won't be merged into 2.0.

**Pull request closed without merging**